### DISCUSSION TITLE
People using these dots-files, how do you preserve your own changes when updating?

### DISCUSSION THREAD

**Discussion by KAGEYAM4** [end-4/dots-hyprland#339] [Q&A]
went through the wiki and there was no mention about preseving own changes.

What are you workflows regrarding this? The only thing i can think of is to fork -> maintain seprate branch -> rebase everytime i want new changes from end-4/dots-hyprland -> use ./install.sh

**Comment by clsty** [comment]
As for myself, refer to https://github.com/clsty/dots-hyprarch
It's not a fork (in terms of git), and I frequently sync the `.config/ags` folder from this repo as it changes often.

Also, the install script will not overwrite `~/.config/ags/user_options.js` and `~/.config/hypr/{hyprland.conf,custom}`.

I maintain the install script so if you have some suggestions I'll be happy to hear them.
Though I'll be VERY busy this month.

**Comment by KAGEYAM4** [comment]
> I frequently sync the `.config/ags` folder from this repo as it changes often.

How do you sync from upstream? `rsync`? Just give me a keyword if the process is lengthy, i will google it.

**Reply by clsty** [reply]
Exactly, just rsync the whole folder first.
Any changes on it can be done by script later, e.g. apply a patch file (keyword: diff/patch) or using sed/awk or just directly copy-paste from another place.

**Reply by Nevrdid** [reply]
Hi, another solution, just write a shell script which will append requires, source or any include line to the  configs file you want to keep modifications and link yours personals modifications. Or, my one, i just commited my changes to my local ~/.cache/dots-hyprland repo and i git pull --rebase when needed.

> I maintain the install script so if you have some suggestions I'll be happy to hear them.

@clsty About a suggestion, why not just create .new files anywhere there are modifications and display it?

**Reply by jaynis** [reply]
@clsty how do you prevent custom changes to the `.config/ags` folder from being overwritten when syncing it with the upstream? Or is that (customizing AGS) simply something which is not intended apart from the `user_options.js`?

**Reply by Nevrdid** [reply]
If you do all you modifications from `~/.cache/dots-hyprland/.config` then `git add .config/ags; git commit -m "My ags changes"; git pull --rebase origin main`  does the job. If you have conflicts, fix them and `rebase --continue`

**Reply by jaynis** [reply]
Yeah I am familiar with git and its capabilities and your suggested workflow will work - until it doesnt. Which means until there are major changes in the upstream and git does not know anymore how to bring them together with the local customizations. Maybe I should have been a bit more clearer in my initial question :smile:. What I actually want to know is: Is there a more elegant solution than potentially having to merge changes by hand on every update?

**Reply by Nevrdid** [reply]
Well... the "until it doesn't" is when conflicts occur and you have to care about the changes. If you just expect to not encounter conflicts, either don't change anything or not update your config ^^'
However, until you change half the configuration, i don't think you'll encounter conflicts anytime you git pull.

**Reply by clsty** [reply]
> @clsty how do you prevent custom changes to the .config/ags folder from being overwritten when syncing it with the upstream? Or is that (customizing AGS) simply something which is not intended apart from the user_options.js?

@jaynis Nope, it's not intended. Precisely, we are not blocking it, we just don't support it for technical reasons (i.e. no good solution yet).

You have to handle it yourself if you want to customizing things other than `user_options.js`. Though you're free to open an issue to suggest more things to be added to `user_options.js`.

As for me, I have a customized cheatsheet according to my customized keybindings. So, I have a script to do the following things in a row (note: `$base` is a path where I store these folders):
1. `cd` to folder `$base/dots-hyprland` and `git pull`,
2. `rsync -av --delete $base/dots-hyprland/.config/ags/ $base/my-dots/.config/ags/`, which means a total copy-and-overwrite.
3. Apply the patch for the cheetsheat data file under `$base/my-dots/.config/ags/`. Not anything else because I don't need to.
4. Run `$base/my-dots/setup.sh -fs` which is the install script similar to `./install.sh -fs` in this repo.

**Comment by clsty** [comment]
> Hi, another solution, just write a shell script which will append requires, source or any include line to the configs file you want to keep modifications and link yours personals modifications. Or, my one, i just commited my changes to my local ~/.cache/dots-hyprland repo and i git pull --rebase when needed.
> 
> > I maintain the install script so if you have some suggestions I'll be happy to hear them.
> 
> @clsty About a suggestion, why not just create .new files anywhere there are modifications and display it?

I'm not a fan of abusing `.new` files, though I guess it could be an optional function. I may consider this in future, though with low priority. Added to my todo list.

**Reply by Nevrdid** [reply]
I have another idea. Why not create a update config script which will look at diff between actuals configs files and the commit which given them( we could keep somewhere a sha of the commit from where the config used is). then backup those files in /tmp or ~/.cache. Next, does the regular rsyncs and ask at the end if the user want to apply diff or just display them to manually restore them.

**Reply by clsty** [reply]
Diff/patch is not robust at all. As for myself, I only use diff/patch for keybinding cheatsheet data for now. Even just with this single file, I still have to fix the patch manually when the applying failed.

**Reply by clsty** [reply]
I already have some idea on the install script but I can't do anything until next month.

**Reply by clsty** [reply]
Ok I've implemented some new options. That's the basic part of the idea I've mentioned above.

Run `./install.sh -h` to see usage.
